OpenDreamClient/Interface/InterfaceMacro.cs

```diff
@@ -39,45 +43,263 @@ public sealed class InterfaceMacroSet : InterfaceElement {
+                    break;
+                case "CTRL":
+                    parsed.Ctrl = true;
+                    break;
+                case "ALT":
+                    parsed.Alt = true;
+                    break;
+                default:
+                    if (part == "ANY") {
+                        parsed.IsAny = true;
+                    } else {
+                        parsed.Key = KeyNameToKey(part);
+                        if (parsed.Key == Keyboard.Key.Unknown) {
+                            throw new Exception($"Invalid keybind part: {part}");
+                        }
+                    }
+
+                    if (foundKey) {
+                        throw new Exception($"Duplicate key in keybind: {part}");
+                    }
+                    foundKey = true;
+                    break;
+            }
+        }
+
+        return parsed;
+    }
+}
+
 public sealed class InterfaceMacro : InterfaceElement {
     public string Id => MacroDescriptor.Id;
     public string Command => MacroDescriptor.Id;
 
     private MacroDescriptor MacroDescriptor => (ElementDescriptor as MacroDescriptor);
 
     private readonly IEntitySystemManager _entitySystemManager;
+    private readonly IUserInterfaceManager _uiManager;
 
     private readonly bool _isRepeating;
     private readonly bool _isRelease;
+    private readonly bool _isAny;
 
-    public InterfaceMacro(string contextName, MacroDescriptor descriptor, IEntitySystemManager entitySystemManager, IInputManager inputManager, IInputCmdContext inputContext) : base(descriptor) {
+    public InterfaceMacro(string contextName, MacroDescriptor descriptor, IEntitySystemManager entitySystemManager, IInputManager inputManager, IInputCmdContext inputContext, IUserInterfaceManager uiManager) : base(descriptor) {
         _entitySystemManager = entitySystemManager;
+        _uiManager = uiManager;
 
         BoundKeyFunction function = new BoundKeyFunction($"{contextName}_{Id}");
-        KeyBindingRegistration binding = CreateMacroBinding(function, Name.ToUpperInvariant());
-
-        if (binding == null)
+        ParsedKeybind parsedKeybind;
+        
+        try {
+            parsedKeybind = ParsedKeybind.Parse(Name);
+        } catch (Exception e) {
+            Logger.Warning($"Invalid keybind for macro {Id}: {e.Message}");
             return;
+        }
+
+        _isRepeating = parsedKeybind.Rep;
+        _isRelease = parsedKeybind.Up;
+        _isAny = parsedKeybind.IsAny;
 
-        _isRepeating = Name.Contains("+REP");
-        _isRelease = Name.Contains("+UP");
+        if (_isAny && (parsedKeybind.Shift || parsedKeybind.Ctrl || parsedKeybind.Alt || parsedKeybind.Rep))
+            throw new Exception("ANY can only be combined with the +UP modifier");
 
         if (_isRepeating && _isRelease)
             throw new Exception("A macro cannot be both +REP and +UP");
 
+        if (_isAny) {
+            inputManager.FirstChanceOnKeyEvent += FirstChanceKeyHandler;
```

Comment:
`FirstChanceKeyHandler` ends up running regardless of which macro set is enabled.